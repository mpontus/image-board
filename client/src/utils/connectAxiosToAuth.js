import { Observable, empty } from "rxjs";
import { switchMap } from "rxjs/operators";
import { set, lensPath } from "ramda";

/**
 * Connects axios instance AuthService using interceptors
 *
 * @param axios Axios
 * @param authService AuthService
 * @return fn Unsubscribe function
 */
const connectAxiosToAuth = (axios, authService) =>
  authService
    .getIdToken()
    .pipe(
      switchMap(
        token =>
          token
            ? Observable.create(() => {
                const interceptor = set(
                  lensPath(["headers", "Authorization"]),
                  `Bearer ${token}`
                );

                axios.interceptors.request.use(interceptor);

                return () => axios.interceptors.request.eject(interceptor);
              })
            : empty()
      )
    )
    .subscribe();
export default connectAxiosToAuth;
