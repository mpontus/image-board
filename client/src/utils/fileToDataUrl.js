/**
 * Convert Blob or File to data url containing Base64 string.
 *
 * @param file File | Blob Blob or file
 * @return Promise<string> Data URL
 */
const fileToDataUrl = file =>
  new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.readAsDataURL(file);
    reader.onload = () => resolve(reader.result);
    reader.onerror = reject;
  });

export default fileToDataUrl;
