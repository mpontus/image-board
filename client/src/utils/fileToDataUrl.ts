/**
 * Convert File to data url
 *
 * @param file File | Blob Blob or file
 * @return Promise<string> Data URL
 */
const fileToDataUrl = (file: File): Promise<string> =>
  new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.readAsDataURL(file);
    reader.onload = () => {
      if (reader.result == null) {
        reject(new Error("Reader result is null"));
      } else if (reader.result instanceof ArrayBuffer) {
        resolve(reader.result.toString());
      } else {
        resolve(reader.result);
      }
    };
    reader.onerror = reject;
  });

export default fileToDataUrl;
