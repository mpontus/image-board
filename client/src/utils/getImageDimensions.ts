interface Dimensions {
  width: number;
  height: number;
}

/**
 * Return dimensions of an image specified as a File
 *
 * @param file File Image file
 * @return Dimensions Object containing width and height
 */
const getImageDimensions = (file: File): Promise<Dimensions> =>
  new Promise((resolve, reject) => {
    const image = new Image();
    image.src = URL.createObjectURL(file);
    image.onload = () =>
      resolve({
        width: image.width,
        height: image.height,
      });
  });

export default getImageDimensions;
