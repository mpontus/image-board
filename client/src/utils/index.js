export { default as fileToDataUrl } from "./fileToDataUrl";
export { default as connectAxiosToAuth } from "./connectAxiosToAuth";
