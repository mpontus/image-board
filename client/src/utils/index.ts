export { default as getImageDimensions } from "./getImageDimensions";
export { default as fileToDataUrl } from "./fileToDataUrl";
