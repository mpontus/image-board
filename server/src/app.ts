import * as cors from "cors";
import * as dotenv from "dotenv-safe";
import * as express from "express";
import * as jwt from "express-jwt";
import * as paginate from "express-paginate";
import * as jwksRsa from "jwks-rsa";
import * as mongoose from "mongoose";
import * as morgan from "morgan";
import * as multer from "multer";
import * as shortid from "shortid";
import * as winston from "winston";
import Post from "./model/Post";
import createStorage from "./storage/createStorage";

dotenv.config();

const storage = createStorage({
  storageType: process.env.CLOUDINARY_STUB ? "test" : "cloudinary",
  cloudinaryUrl: process.env.CLOUDINARY_URL,
  storageParams: {
    folder: process.env.CLOUDINARY_FOLDER || "",
    allowedFormats: ["jpg", "png"],
    filename: (req, file, cb) => cb(null, shortid())
  }
});

mongoose.connect(
  process.env.MONGODB_URI || "",
  {
    useNewUrlParser: true
  }
);

const upload = multer({
  storage
});

const logger = winston.createLogger({
  transports: [new winston.transports.Console()]
});

const app = express();

app.use(morgan("combined"));
app.use(cors());

app.use(
  jwt(
    process.env.NODE_ENV === "test"
      ? {
          secret: process.env.JWT_SECRET || "",
          credentialsRequired: false
        }
      : {
          secret: jwksRsa.expressJwtSecret({
            cache: true,
            rateLimit: true,
            jwksRequestsPerMinute: 5,
            jwksUri: `https://${process.env.AUTH0_DOMAIN}/.well-known/jwks.json`
          }),
          audience: process.env.AUTH0_CLIENT_ID,
          issuer: `https://${process.env.AUTH0_DOMAIN}/`,
          algorithms: ["RS256"],
          credentialsRequired: false
        }
  )
);

const requireAuth: express.RequestHandler = (req, res, next) => {
  if (!req.user) {
    res.status(401).end();
  } else {
    next();
  }
};

interface User {
  sub: string;
  nickname: string;
  picture: string;
}

type ModelType<T> = T extends mongoose.Model<infer R> ? R : any;

const serializePost = (user: User) => (post: ModelType<typeof Post>) => ({
  id: post._id,
  imageUrl: post.imageUrl,
  imageWidth: post.imageWidth,
  imageHeight: post.imageHeight,
  author: {
    id: post.author.id,
    name: post.author.name,
    avatarUrl: post.author.avatarUrl
  },
  likes: post.likesCount,
  isLiked: user ? post.likes.includes(user.sub) : false,
  timestamp: post.timestamp
});

app.get("/api/posts", paginate.middleware(5, 5), async (req, res, next) => {
  try {
    const [total, posts] = await Promise.all([
      Post.find({})
        .countDocuments()
        .exec(),

      Post.find({})
        .sort("-timestamp")
        .skip(req.skip as number)
        .limit(req.query.limit)
        .lean()
        .exec()
    ]);

    res.json({
      total,
      items: posts.map(serializePost(req.user))
    });
  } catch (error) {
    next(error);
  }
});

app.post(
  "/api/posts",
  requireAuth,
  upload.single("file"),
  async (req, res, next) => {
    try {
      if (!req.user.nickname || !req.user.picture) {
        res.status(400).send();
      }

      const file = req.file as any;

      const post = await Post.create({
        imageId: file.public_id,
        imageUrl: file.url,
        imageWidth: file.width,
        imageHeight: file.height,
        author: {
          id: req.user.sub,
          name: req.user.nickname,
          avatarUrl: req.user.picture
        },
        likes: [req.user.sub],
        likesCount: 1
      });

      res.status(201).json(serializePost(req.user)(post));
    } catch (error) {
      next(error);
    }
  }
);

app.delete("/api/posts/:id", requireAuth, async (req, res, next) => {
  try {
    const post = mongoose.Types.ObjectId.isValid(req.params.id)
      ? await Post.findOne({ _id: req.params.id })
      : null;

    if (post === null) {
      res.status(404).send();

      return;
    }

    if (req.user.sub !== post.author.id) {
      res.status(401).send();
    }

    await Promise.all([storage.delete(post.imageId), post.remove()]);

    res.status(204).end();
  } catch (error) {
    next(error);
  }
});

app.put("/api/posts/:id/like", requireAuth, async (req, res, next) => {
  try {
    const { nModified } = await Post.updateOne(
      {
        _id: req.params.id,
        likes: { $ne: req.user.sub }
      },
      {
        $inc: {
          likesCount: 1
        },
        $push: {
          likes: req.user.sub
        }
      }
    ).exec();

    if (nModified === 0) {
      res.status(400).end();

      return;
    }

    res.status(202).end();
  } catch (error) {
    next(error);
  }
});

app.delete("/api/posts/:id/like", requireAuth, async (req, res, next) => {
  try {
    const { nModified } = await Post.updateOne(
      {
        _id: req.params.id,
        likes: req.user.sub
      },
      {
        $inc: {
          likesCount: -1
        },
        $pull: {
          likes: req.user.sub
        }
      }
    ).exec();

    if (nModified === 0) {
      res.status(400).end();

      return;
    }

    res.status(202).end();
  } catch (error) {
    next(error);
  }
});

app.use(((err, req, res, next) => {
  logger.error(err);

  res.status(500).json({
    message: err.message
  });
}) as express.ErrorRequestHandler);

export default app;
