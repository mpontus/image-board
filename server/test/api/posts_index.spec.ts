import app from "@src/app";
import * as request from "supertest";
import { createToken, resetDb } from "../utils";

beforeEach(() => resetDb());

describe("/api/posts", () => {
  describe("GET", () => {
    describe("when no posts exist", () => {
      it("should return a valid response", async () => {
        const response = await request(app)
          .get("/api/posts")
          .expect(200);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe("when some posts exist", () => {
      beforeEach(() => require("@test/seed/single-post").default());

      it("should return a valid response", async () => {
        const response = await request(app)
          .get("/api/posts")
          .expect(200);

        expect(response.body).toMatchSnapshot();
      });
    });
  });

  describe("POST", () => {
    describe("when file is an image", () => {
      it("should return a valid response", async () => {
        const token = createToken({
          sub: "123",
          nickname: "Alex Pushkin",
          picture: "https://picsum.photos/200/200"
        });

        const response = await request(app)
          .post("/api/posts")
          .attach("file", __dirname + "/../fixtures/sample.jpg")
          .set("Authorization", `Bearer ${token}`)
          .expect(201);

        expect(response.body).toMatchSnapshot({
          id: expect.any(String),
          timestamp: expect.any(String)
        });
      });
    });

    describe("when token contains insufficient information", () => {
      it("should return a valid response", async () => {
        const token = createToken({
          nickname: undefined
        });

        const response = await request(app)
          .post("/api/posts")
          .attach("file", __dirname + "/../fixtures/sample.jpg")
          .set("Authorization", `Bearer ${token}`)
          .expect(400);

        expect(response.body).toMatchSnapshot();
      });
    });
  });
});
