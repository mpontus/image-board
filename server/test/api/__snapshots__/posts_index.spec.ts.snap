// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/api/posts GET when no posts exist should return a valid response 1`] = `
Object {
  "items": Array [],
  "total": 0,
}
`;

exports[`/api/posts GET when some posts exist should return a valid response 1`] = `
Object {
  "items": Array [
    Object {
      "author": Object {
        "avatarUrl": "http://lorempixel.com/640/480",
        "id": "7862f3cc-bfc5-41b8-aded-1d0420ea196a",
        "name": "Vernie.Jenkins",
      },
      "id": "43345823304969c878318d12",
      "imageHeight": 417,
      "imageUrl": "http://lorempixel.com/640/480",
      "imageWidth": 914,
      "isLiked": false,
      "likes": 0,
      "timestamp": "1986-12-02T10:15:00.816Z",
    },
  ],
  "total": 1,
}
`;

exports[`/api/posts POST when file is an image should return a valid response 1`] = `
Object {
  "author": Object {
    "avatarUrl": "https://picsum.photos/200/200",
    "id": "123",
    "name": "Alex Pushkin",
  },
  "id": Any<String>,
  "imageHeight": 1306,
  "imageUrl": "http://res.cloudinary.com/du0jftvmw/image/upload/v1537197602/image-board/NBsz2e5Wv.jpg",
  "imageWidth": 1000,
  "isLiked": true,
  "likes": 1,
  "timestamp": Any<String>,
}
`;

exports[`/api/posts POST when token contains insufficient information should return a valid response 1`] = `Object {}`;
